nemo_skills/inference/prover.py

```diff
@@ -0,0 +1,428 @@
+                "nemotron_refinement_prompt_config is required when nemotron_refinement is enabled."
+            )
+            self.nemotron_refine_prompt = get_prompt(self.cfg.nemotron_refinement_prompt_config)
+
+    # with adaptive reasoning
+    async def _generate_single_completion(self, prompt: str, **kwargs):
+        if is_dataclass(self.cfg.inference):
+            inference_params = asdict(self.cfg.inference)
+        else:
+            # Already a dict from Hydra
+            inference_params = dict(self.cfg.inference)
+        generation_params = {
+            "prompt": prompt,
+            "stop_phrases": [self.cfg.stop_phrase] if self.cfg.stop_phrase else None,
+            **inference_params,
+            **self.extra_generate_params,
+        }
+        # Override endpoint_type to text since we already applied the chat template
+        generation_params["endpoint_type"] = EndpointType.text
+        for key, value in kwargs.items():
+            generation_params[key] = value
+        generation = await self.llm.generate_async(**generation_params)
+        if self.cfg.adaptive_reasoning:
+            assert generation_params["extra_body"].get("reasoning_effort", None) is not None, (
+                "reasoning_effort is required when adaptive_reasoning is enabled"
+            )
+            reasoning_effort_index = reasoning_effort_list.index(
+                generation_params["extra_body"].get("reasoning_effort", None)
+            )
+            while len(generation["generation"]) == 0 and reasoning_effort_index > 0:
+                LOG.info(
+                    "Reasoning effort is too high, reducing to %s", reasoning_effort_list[reasoning_effort_index - 1]
+                )
+                reasoning_effort_index = reasoning_effort_index - 1
+                generation_params["extra_body"]["reasoning_effort"] = reasoning_effort_list[reasoning_effort_index]
+                generation = await self.llm.generate_async(**generation_params)
+        if self.cfg.parse_generation:
+            parse_reasoning(
+                generation,
+                self.cfg.generation_key,
+                self.cfg.end_reasoning_string,
+            )
+        return generation
+
+    # factor out this part so it won't become a bottleneck.
+    async def _extract_and_replace_code(self, formal_statement, generation):
+        code = extract_code(generation)
+        full_code = replace_statement_in_proof(formal_statement, code)
+        return code, full_code
+
+    def _transform_for_nemotron_refinement(self, proof_attempt: str, error_message: str) -> list[dict]:
+        """Transform multi-turn refinement into single-turn nemotron-style prompt."""
+        return self.nemotron_refine_prompt.fill(
+            {
+                "proof_attempt": proof_attempt,
+                "error_message": error_message,
+            }
+        )
+
+    async def _single_data_point_generate(self, data_point, data):
```

Comment:
per this comment, wrap this whole method in `with self.semaphore`: https://github.com/NVIDIA-NeMo/Skills/pull/1078/files#r2593972374

You can make a helper to not have to indent this whole block if you want.